dig.go

```diff
@@ -0,0 +1,277 @@
+	}
+	args, err := c.constructorArgs(ftype)
+	if err != nil {
+		return fmt.Errorf("failed to get arguments for %v (type %v): %v", function, ftype, err)
+	}
+	returned := reflect.ValueOf(function).Call(args)
+	if len(returned) == 0 {
+		return nil
+	}
+	if last := returned[len(returned)-1]; last.Type() == _errType && last.Interface() != nil {
+		return fmt.Errorf("failed to execute %v (type %v): %v", function, ftype, last.Interface())
+	}
+	return nil
+}
+
+func (c *Container) provideInstance(val interface{}) error {
+	vtype := reflect.TypeOf(val)
+	if vtype == _errType {
+		return errors.New("can't provide errors")
+	}
+	if _, ok := c.nodes[vtype]; ok {
+		return errors.New("already in container")
+	}
+	c.nodes[vtype] = node{provides: vtype}
+	c.cache[vtype] = reflect.ValueOf(val)
+	return nil
+}
+
+func (c *Container) provideConstructor(ctor interface{}, ctype reflect.Type) error {
+	returnTypes := make(map[reflect.Type]struct{}, ctype.NumOut())
+	for i := 0; i < ctype.NumOut(); i++ {
+		rt := ctype.Out(i)
+		if rt == _errType {
+			// Don't register errors into the container.
+			continue
+		}
+		if _, ok := returnTypes[rt]; ok {
+			return fmt.Errorf("returns multiple %v", rt)
+		}
+		if _, ok := c.nodes[rt]; ok {
+			return fmt.Errorf("provides type %v, which is already in the container", rt)
+		}
+		returnTypes[rt] = struct{}{}
+	}
+	if len(returnTypes) == 0 {
+		return errors.New("must provide at least one non-error type")
+	}
+
+	nodes := make([]node, 0, len(returnTypes))
+	for rt := range returnTypes {
+		n := newNode(rt, ctor, ctype)
+		nodes = append(nodes, n)
+		c.nodes[rt] = n
+	}
+
+	for _, n := range nodes {
+		if err := c.isAcyclic(n); err != nil {
+			c.remove(nodes)
+			return fmt.Errorf("introduces a cycle: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Container) isAcyclic(n node) error {
+	return detectCycles(n, c.nodes, nil, make(map[reflect.Type]struct{}))
+}
+
+func (c *Container) get(t reflect.Type) (reflect.Value, error) {
+	if v, ok := c.cache[t]; ok {
+		return v, nil
+	}
+
+	n, ok := c.nodes[t]
+	if !ok {
+		return _noValue, fmt.Errorf("type %v isn't in the container", t)
+	}
+
+	if err := c.contains(n.deps); err != nil {
+		return _noValue, fmt.Errorf("missing dependencies for type %v: %v", t, err)
+	}
+
+	args, err := c.constructorArgs(n.ctype)
+	if err != nil {
+		return _noValue, fmt.Errorf("couldn't get arguments for constructor %v: %v", n.ctype, err)
+	}
+	constructed := reflect.ValueOf(n.ctor).Call(args)
+
+	// Provide-time validation ensures that all constructors return at least
+	// one value.
+	if err := constructed[len(constructed)-1]; err.Type() == _errType && err.Interface() != nil {
+		return _noValue, fmt.Errorf("constructor %v for type %v failed: %v", n.ctype, t, err.Interface())
+	}
+
+	for _, con := range constructed {
+		ct := con.Type()
+		if ct == _errType {
+			continue
+		}
+		c.cache[ct] = con
+	}
+	return c.cache[t], nil
+}
+
+func (c *Container) contains(types []reflect.Type) error {
```

Comment:
this can be part of `constructorArgs` just having a `contains` method would lead to its misuse.